Added menubar item to start and stop timers.

marcjulianschwarz opened the pull request:
Hi @ThatNerdSquared,

added a new menubar command that will allow users to manage timers from their menubar. Like this:

<img width="451" alt="Bildschirmfoto 2022-07-29 um 11 00 33" src="https://user-images.githubusercontent.com/67844154/181724158-68a1ba41-2e13-4c65-82ab-886cf6163369.png">

With this change I updated the Raycast API to its newest version and removed all `await`s and `async`s as I couldn't find any method being invoked asynchronously.

marcjulianschwarz commented:
Hmm, looks like there are still some issues when clicks on "Start x Minute Timer" or "Stop Timer" wont register.
Dont know where the problem is.

ThatNerdSquared commented:
Thanks for the PR! I'll look at this tomorrow, need to catch up on the new Raycast API changes and test your changes out locally.

ThatNerdSquared commented:
Hmm... the MenuBarExtra doesn't seem to show up when there are no timers running. Also, after starting a timer, and then stopping it via the MenuBarExtra, the MenuBarExtra doesn't seem to update to reflect that there are currently no timers running. Can you reproduce this?
Apologies if these are problems on my end - I haven't dug into the MenuBarExtra API very much yet.

marcjulianschwarz commented:
Yes noticed this too now. Looks like it is because of the `isLoading` prop which is only set to false when a timer exists. This was to prevent empty state flickering. But there must be a way to deal with this. 

The second issue you mentioned is the same that I was talking about here:

> Hmm, looks like there are still some issues when clicks on "Start x Minute Timer" or "Stop Timer" wont register.
> Dont know where the problem is.

A lot of times the MenuBarItem won't register a click which is weird. I have added some `console.log`s to see when the click is actually registered and somehow this doesn't happen often.

marcjulianschwarz commented:
Hi @ThatNerdSquared,

if you dont mind you can try out the new commit. The starting and stopping of timers in the menubar should now work as expected. And the menubar will be visible even without any timers running in the beginning.

The issue with stopping and starting timers was probably on raycasts side as I updated their API and now everything is working. See [their changelog](https://developers.raycast.com/changelog).

marcjulianschwarz commented:
https://user-images.githubusercontent.com/67844154/183747860-c519fc7e-1a4e-4347-ab9a-d17c29af5a74.mov

ThatNerdSquared commented on src/timerUtils.ts:
```diff
@@ -62,44 +67,69 @@ async function getTimers() {
   writeFileSync(DATAPATH, JSON.stringify(customTimers));
 }
 
+function formatTime(seconds: number) {
```
Out of curiosity... I understand extracting this out to the `timerUtils`, but why reimplement it?

marcjulianschwarz replied:
I had some issues where it would add an hour because of my timezone. 

However I just tried the old implementation and it works as expected now. Maybe I had something wrong in the beginning.
Should I keep the current implementation or change it to yours?

I prefer yours as it is way shorter.

ThatNerdSquared commented on src/manageTimers.tsx:
```diff
@@ -19,47 +20,41 @@ export default function Command() {
-    setInterval(async () => {
-      await refreshTimers();
+    setInterval(() => {
+      setTimers(getTimers());
```
You should call `refreshTimers` here to load custom timers as well - your current approach only loads running timers, not presets.

ThatNerdSquared commented on src/timersMenuBar.tsx:
```diff
@@ -0,0 +1,95 @@
+    startTimer(seconds, name);
+  }
+
+  console.log(timers);
```
Assuming this and a few other `console.log`s are left over from debugging, they're no longer needed.

ThatNerdSquared reviewed (changes requested):
Looks good! I'll merge the PR as soon as you respond to a few small fixes/questions below. Feel free to create a `contributors` field in the `package.json` and add yourself to it as mentioned in the [Raycast manifest docs](https://developers.raycast.com/information/manifest) as well.

marcjulianschwarz commented:
Will add these changes later this day. Thanks for reviewing.

marcjulianschwarz commented:
I reverted to the old `formatTime` function as I can't reproduce the issue that I had with it. I removed all debugging artifacts and added the `refreshTimers` call.

ThatNerdSquared reviewed (approved):
Looks good! I have a few fixes I'm working on locally, after I finish those up, I'll push those along with your changes to the Store. Thanks for your PR!

marcjulianschwarz commented:
No problem. Always wanted to have timers in the menubar but didn’t want to install a whole new app just for this one purpose.

Your extension made that possible :)
